wallet/update.go

```diff
@@ -36,98 +45,207 @@ type (
-	index := cau.State.Index
-	state := cau.State
+// appliedEvents returns a slice of events that are relevant to the wallet
+// in the chain update.
+func appliedEvents(cau chain.ApplyUpdate, walletAddress types.Address) (events []Event) {
+	cs := cau.State
 	block := cau.Block
+	index := cs.Index
+	siacoinElements := make(map[types.SiacoinOutputID]types.SiacoinElement)
 
-	// cache the source of new immature outputs to show payout transactions
-	if state.FoundationPrimaryAddress == addr {
-		events = append(events, Event{
-			ID:    types.Hash256(index.ID.FoundationOutputID()),
-			Index: index,
-			Transaction: types.Transaction{
-				SiacoinOutputs: []types.SiacoinOutput{
-					state.FoundationSubsidy(),
-				},
-			},
-			Inflow:    state.FoundationSubsidy().Value,
-			Source:    EventSourceFoundationPayout,
+	// cache the value of spent siacoin elements to use when calculating outflow
+	cau.ForEachSiacoinElement(func(se types.SiacoinElement, spent bool) {
+		if spent {
+			siacoinElements[types.SiacoinOutputID(se.ID)] = se
+		}
+	})
+
+	addEvent := func(id types.Hash256, data any) {
+		ev := Event{
+			ID:        id,
+			Index:     index,
+			Data:      data,
 			Timestamp: block.Timestamp,
-		})
+		}
+
+		switch data := data.(type) {
+		case EventMinerPayout:
+			ev.Inflow = data.SiacoinElement.SiacoinOutput.Value
+			ev.Type = EventTypeMinerPayout
+			ev.MaturityHeight = cs.MaturityHeight()
+		case EventFoundationSubsidy:
+			ev.Inflow = data.SiacoinElement.SiacoinOutput.Value
+			ev.Type = EventTypeFoundationSubsidy
+			ev.MaturityHeight = cs.MaturityHeight()
+		case types.Transaction:
+			for _, si := range data.SiacoinInputs {
+				if si.UnlockConditions.UnlockHash() == walletAddress {
+					ev.Outflow = ev.Outflow.Add(siacoinElements[si.ParentID].SiacoinOutput.Value)
+				}
+			}
+
+			for _, so := range data.SiacoinOutputs {
+				if so.Address == walletAddress {
+					ev.Inflow = ev.Inflow.Add(so.Value)
+				}
+			}
+			ev.MaturityHeight = index.Height
+			ev.Type = EventTypeV1Transaction
+		case EventV1ContractPayout:
+			ev.Inflow = data.SiacoinElement.SiacoinOutput.Value
+			ev.Type = EventTypeV1Contract
+			ev.MaturityHeight = cs.MaturityHeight()
+		case types.V2Transaction:
+			for _, si := range data.SiacoinInputs {
+				if si.SatisfiedPolicy.Policy.Address() == walletAddress {
+					ev.Outflow = ev.Outflow.Add(siacoinElements[types.SiacoinOutputID(si.Parent.ID)].SiacoinOutput.Value)
+				}
+			}
+
+			for _, so := range data.SiacoinOutputs {
+				if so.Address == walletAddress {
+					ev.Inflow = ev.Inflow.Add(so.Value)
+				}
+			}
+			ev.Type = EventTypeV2Transaction
+			ev.MaturityHeight = index.Height
+		case EventV2ContractPayout:
+			ev.Inflow = data.SiacoinElement.SiacoinOutput.Value
+			ev.Type = EventTypeV2Contract
+			ev.MaturityHeight = cs.MaturityHeight()
+		}
+
+		events = append(events, ev)
+	}
+
+	relevantV1Txn := func(txn types.Transaction) bool {
```

Comment:
wondering whether it's worth adding a `Contains` on some of these types for which we keep writing `isRelevantX` for

Reply:
I had the same thought. Obviously it would need to be in `core` but we essentially have the same code in every piece of node software.

Reply:
I don't think it's necessary. The logic is simple and there may be optimizations for exiting early that would be ignored if we added a helper.